### ISSUE TITLE
How to model a Siamese Network?

### ISSUE THREAD

**Issue by zgxiangyang** [BVLC/caffe#316]
which is usefull to solve a two classification problems.

[8] S. Chopra, R. Hadsell, and Y. LeCun. Learning a similarity metric discriminatively, with application to face veriﬁcation. In CVPR, 2005. 2, 5

**Comment by anxiaoxi45** [comment]
@sunbaigui hi,Have you solved the problem? I also want to know how write the image list file, did it work like this format (img1 img2 label) ? How did you do?

**Comment by sunbaigui** [comment]
@anxiaoxi45 I havn't done this yet, sorry for reply so late, kind of busy recently, have you done it yet?

**Comment by gfl94** [comment]
@anxiaoxi45 @sunbaigui Have you solved the  (img1 img2 label) problem?

**Comment by szahid86** [comment]
@gfl94 : I have tried to solve the (img1 img2 label) problem using following approach but failed. Please note that I am very newbie to python and caffe. I followed a very pingpong approach.

1- I decided to use the Siamese example file (mnist_siamese_train_test.prototxt), and make some changes in it. The example mnist_siamese_train_test.prototxt requires the input train and test data to be in lmdb format.

2- So I tried to convert my train and test data into lmdb format using convert_mnist_siamese_data.cpp but I was unable to convert my data because of some weird errors.

3- Next I found some code, that shows the possibility to use .jpg files at data layer, rather than converting to lmdb format. That code uses two train1.txt and train2.txt files (with the format, path/to/image/1.jpg label). Label value is '1' if the image files in the corresponding lines of the two .txt files are of the same person, and '0' otherwise.

Sample Data layers for your reference:

layers {
  name: "data"
  type: IMAGE_DATA
  top: "data"
  top: "label"
  image_data_param {
    source: "/home/cvip/caffe/examples/siamese/My_DB/TrainLabels_1.txt"
    batch_size: 128
  }
  transform_param {
 # crop_size: 40
  mirror: true
  }
  include: { phase: TRAIN }
}
layers {
  name: "data"
  type: IMAGE_DATA
  top: "data"
  top: "label"
  image_data_param {
    source: "/home/cvip/caffe/examples/siamese/lfw_funneled/TestImageLabels_1.txt"
    batch_size: 100
  }
  transform_param {
 # crop_size: 40
  mirror: true

  }
  include: { phase: TEST }
}

layers {
  name: "data_p"
  type: IMAGE_DATA
  top: "data_p"
  top: "label_p"
  image_data_param {
    source: "/home/cvip/caffe/examples/siamese/My_DB/TrainLabels_2.txt"
    batch_size: 128
  }
  transform_param {
 # crop_size: 40
  mirror: true

  }
  include: { phase: TRAIN }
}
layers {
  name: "data_p"
  type: IMAGE_DATA
  top: "data_p"
  top: "label_p"
  image_data_param {
    source: "/home/cvip/caffe/examples/siamese/lfw_funneled/TestImageLabels_2.txt"
    batch_size: 100
  }
  transform_param {
 # crop_size: 40
  mirror: true
  }
  include: { phase: TEST }
}


4- Using this code I was able to start the training of the Siamese network, but the training was not converging.

5- I made some changes in the prototxt file (changed some filters and parameters in the convolutional layer) and the solver (reduced the learning rate), and was able to run the code without divergence.

6- The problem I am facing right now is that the loss value (during the training steps) is making no sense to me.. the values ranges from 0.56 ~ 0.0019, and there is no pattern of occurrence of these values.

Please share your experience of using Siamese network for this kind of problem (face verification).

thanks

**Comment by gfl94** [comment]
@szahid86 Thanks for your reply, I'm a newbee for caffe. I tried a familiar thing with you, create the dataset input for siamese network referring to [convert_mnist_siamese_data.cpp](https://github.com/BVLC/caffe/blob/master/examples/siamese/convert_mnist_siamese_data.cpp), generate a datum with a pair of pictures, and write a siamese network with [bvlc_alexnet](https://github.com/BVLC/caffe/tree/master/models/bvlc_alexnet), the result is just not converging, and I am trying to solve this problem.

**Comment by szahid86** [comment]
Hi.. can you share the format of your input_image_file and input_label_file?

I have tried doing this:

input_image_file :- path/to/file/Bob1.jpg path/to/file/Bob2.jpg
input_label_file :- 1

but it gave error ..
I did this a long time ago, so I kind a forget about what the error was..

moreover, I think the input_image_file and input_label_file need to be in some "ubyte" format. So how you converted your database to "ubyte" format?

can you please explain in details?

Thanks

**Comment by camille-jing** [comment]
@szahid86 hi ,How about  your results,  can you tell me the steps in detail how to  make my own  picture  into the siamese net ,thank you !

**Comment by szahid86** [comment]
@camille-jing Hi, 
1) How about my results?
Well I can not compute the accuracy of my trained network because there is not accuracy layer for the type of siamese network I am using. My loss_layer gives a varying output (0.59 ~ 0.001), so I am still not sure that my training routine is working fine or not.
I have switched to Matcaffe interface, which is looking more easy to me. I am planing to extract the features from my last layers and then use the euclidean distance to measure the accuracy manually.

2) Steps in Detail how to make my own picture into the siamese net:
I could not clearly understand what do you mean by "make my own picture" ??
I guess you want to pass your own pictures to the siamese network and compute the output? If this is the case then I would recommend to go through this : http://caffe.berkeleyvision.org/tutorial/interfaces.html

..more specifically you can use the command:
caffe test -model path/to/your/network_train_test.prototxt -weights path/to/your/learnt/model.caffemodel -gpu 0 -iterations 100

This command will use the test image set you provided in the "DATA LAYER" of your network_train_test.prototxt file. If you are using the siamese network with contrastive loss function (same as me) then you should get the output in terms of loss..

Best Regards,

**Comment by camille-jing** [comment]
@szahid86 hi! thanks for your reply!My question is very simple for you ,i think. because my bad-base, so now I don't know how to convert my JPEG images downloaded on the internet  to  the leveldb what Siamese net need format, as it training mnist used command:  / examples/Siamese create_mnist_siamese. Sh, then can be used as input into the network, Now I'm not going to do that.

**Comment by szahid86** [comment]
You can use JPEG files directly, other than using lmdb converted database. Please see my comments above. I have given the sample code of my input data layers. The data layer reads a text file (TrainLabels_1.txt and TrainLabels_2.txt), that points to the JPEG images.
The format of these labels files is as follows:


TrainLabels_1.txt                                                     TrainLabels_1.txt

path/to/img/Bob_1.jpg 1                                           path/to/img/Bob_2.jpg 1
path/to/img/Bob_2.jpg 0                                           path/to/img/Alice_2.jpg 0
path/to/img/Alice_1.jpg 1                                          path/to/img/Alice_2.jpg 1
path/to/img/Alice_1.jpg 0                                          path/to/img/Bob_1.jpg 0

Best Regards,

**Comment by camille-jing** [comment]
@szahid86  Very grateful for your careful answer, I try to use your approach, but when I do in the way you suggested, such as input data for Bob_1. JPG 1 and Bob_1. JPG 0, two data two labels, but  what's the final output ?, how to calculate the  loss, should not the pair of pictures to have only one label, to see if similar?

**Comment by cheer37** [comment]
@szahid86 
hi, did you train your siamese network successfully?
I want to see your rest network in train_test.prototxt
I guess you fed data and data_p blobs into different nets separately.
then two nets would be trained with different parameters. but as i know, in siamese network, two nets share the weights.
so we need the weights for only one network.
What do you think about this?
Thanks.

**Comment by mollyStark** [comment]
@szahid86  Hi~ I'm wondering if you have solved your problem, I ran into similar problem as you, the loss didn't converge, and when I predict the last layer output(e.g 'fc7') of different test images, I got same results... I don't know why, I'll be thankful if you point me a light.

**Comment by szahid86** [comment]
Hi,
Sorry for such a late reply. I was actually working on another project.Well I found some simple matlab code to convert my image data into HDF5 format. I changed the input data layers of my network and trained it again. At this time the training is running, but I am happy that the loss is showing some convergence. Once the training is over, I will compute the results again and let you know if I get any improvements.
If you have already got good results then please share your technique! 
RegardsZahid

**Comment by dawnos** [comment]
@szahid86  Hi~ I met the same problem too, that the siamese network do not converge. Can you share some details?

**Comment by pnambiar** [comment]
I am trying to initialize siamese network with the places data set and then fine-tune with my own custom dataset. I would like to have CNN on both the streams pre-trained on places dataset. How do I accomplish this ?
